src/QsCompiler/DataStructures/DataTypes.fs

```diff
@@ -53,61 +53,118 @@ type QsNullable<'T> = // to avoid having to include the F# core in the C# part o
         | Some v -> Value v
         | None -> Null
 
+/// Operations for QsNullable<'T>.
+module QsNullable =
+    /// Applies 'f' to both nullable arguments if both have a Value. Returns Null otherwise.
+    let Map2 f a b =
+        match a, b with
+        | Value a, Value b -> f a b |> Value
+        | _ -> Null
 
 [<Struct>]
 type NonNullable<'T> = private Item of 'T with
     static member New value = if value = null then ArgumentNullException() |> raise else Item value
     member this.Value = this |> function | Item str -> str
 
 
-[<Struct>]
-type QsPositionInfo = {Line : int; Column: int} with 
-    static member Zero = {Line = 1; Column = 1}
-    static member CombinedRange (left : QsRangeInfo, right : QsRangeInfo) = 
-        match left, right with
-        | Value r1, Value r2 -> Value (fst r1, snd r2)
-        | _ -> Null
+/// A position in a text document.
+type Position = private Position of int * int with
+
+    /// The line number, where line zero is the first line in the document.
+    member this.Line = match this with Position (line, _) -> line
+
+    /// The column number, where column zero is the first column in the line.
+    member this.Column = match this with Position (_, column) -> column
+
+    /// Translates the first position by the amount of the second position. If the resulting position is on the same
+    /// line, the column numbers are added; otherwise, the second position's column number is used.
+    static member (+) (a : Position, b : Position) =
```

Comment:
This seems like a confusing definition of addition. Why isn't it just `(x1, y1) + (x2, y2) = (x1 + x2, y1 + y2)`? Why does the behavior have to change when the line number of the second argument is zero?

Reply:
Okay so the logic has to take the second position's col when when dealing with multiple lines because the col is reset every new line. But when there are no new lines, the col is relative instead of absolute so they need to be added in that case. This all adds up for me 😄.